### ISSUE TITLE
[sdk/nodejs, multi-language components] Cannot read properties of undefined (reading 'invoke') error when calling azure-native

### ISSUE THREAD

**Issue by 4ux-nbIx** [pulumi/pulumi#17074]
### What happened?

We are building a component package in TypeScript that is using azure-native provider inside. The package implements some Azure Resource recipes that we commonly deploy across various stacks in various languages in the organization.

After upgrading to the Pulumi from version 1.121.0 to 1.130.0 calls to azure-native started failing with the following error:

```
TypeError: Cannot read properties of undefined (reading 'invoke'): TypeError: Cannot read properties of undefined (reading 'invoke')

    at C:\snapshot\pulumi-resource-azure-relay\node_modules\@pulumi\azure-native\node_modules\@pulumi\pulumi\runtime\invoke.js:151:114

    at new Promise (<anonymous>)

    at C:\snapshot\pulumi-resource-azure-relay\node_modules\@pulumi\azure-native\node_modules\@pulumi\pulumi\runtime\invoke.js:151:63

    at Generator.next (<anonymous>)

    at fulfilled (C:\snapshot\pulumi-resource-azure-relay\node_modules\@pulumi\azure-native\node_modules\@pulumi\pulumi\runtime\invoke.js:18:58)

    at processTicksAndRejections (node:internal/process/task_queues:95:5)
```

Package.json
``` JSON
    "dependencies": {
        "@pulumi/azure-native": "2.57.1",
        "@pulumi/pulumi": "3.130.0"
    }
```

### Example

``` TypeScript
import * as pulumi from "@pulumi/pulumi";
import * as provider from "@pulumi/pulumi/provider";
import * as azAuth from '@pulumi/azure-native/authorization'
import { ResourceGroup } from "@pulumi/azure-native/resources";


export class Provider implements provider.Provider {
    constructor(readonly version: string, readonly schema: string) { }

    async construct(
        name: string,
        type: string,
        inputs: pulumi.Inputs,
        options: pulumi.ComponentResourceOptions
    ): Promise<provider.ConstructResult> {

        // Throws 
        const azureClientConfig = await azAuth.getClientConfig();
        
        // Works
        new ResourceGroup("test", {resourceGroupName: "test", location: "westeurope"});

        const azureConfig = new pulumi.Config("azure-native");
        // Works
        azureConfig.require("location");

        const myConfig = new pulumi.Config("azure-relay");
        // Works
        myConfig.require("configKey");
    }

    async invoke(token: string, inputs: any): Promise<pulumi.provider.InvokeResult> {

        // Throws 
        const azureClientConfig = await azAuth.getClientConfig();

        const azureConfig = new pulumi.Config("azure-native");
        // Throws
        azureConfig.require("location");

        const myConfig = new pulumi.Config("azure-relay");
        // Throws
        myConfig.require("configKey");
    }
}
```

### Output of `pulumi about`

CLI
Version      3.115.0
Go Version   go1.22.2
Go Compiler  gc

Plugins
KIND      NAME          VERSION
resource  azure-native  2.58.0
resource  azure-relay   1.0.0
language  dotnet        unknown
resource  random        4.16.2

Host
OS       Microsoft Windows 10 Enterprise
Version  10.0.19045 Build 19045
Arch     x86_64

This project is written in dotnet: executable='C:\Program Files\dotnet\dotnet.exe' version='8.0.204'

Backend
Name           ***
URL            azblob://state
User           ***
Organizations
Token type     personal

Dependencies:
NAME                       VERSION
Pulumi                     3.66.2
Pulumi.AzureNative         2.58.0
Pulumi.AzureRelay          1.0.0

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by 4ux-nbIx** [comment]
Update. Checked various pulumi versions and it looks like it broke since version 3.122.0. Version 3.121.0 works as expected.

It also looks like it's only the provider functions in azure-native that stopped working. If our custom provider only created azure-native resources it works without any issues.

Same thing if you call azure-native provider function inside custom provider invoke call.

Also, I noticed that provider functions are getting weird config! I.e. construct calls are getting all the config keys from the parent stack. Invokes only get config keys for the function's provider but with shortened keys.

Example:
construct call receives 
- 'azure-native:location'
- 'azure-relay:configKey'

invoke call receives
- 'configKey'

**Comment by Frassle** [comment]
> Version 3.121.0 works as expected.

Is this just changing the node library version, the CLI is still at 3.115?

**Comment by 4ux-nbIx** [comment]
@Frassle I have updated the code example at the top to beter describe what is working and what is broken.

**Comment by Frassle** [comment]
And I take it in 3.121 all of the examples work?

**Comment by 4ux-nbIx** [comment]
Everything inside construct does work with 3.121, yes! 

When it comes to invokes it's been broken for longer I guess. We've just started working on exposing functions from our provider. I tried the provided examples with 3.121. It failed. I upgraded to the latest hoping it's fixed just to find that everything is broken now.

**Comment by 4ux-nbIx** [comment]
> The behavior you're seeing in `invoke` is by-design based on how the system currently works. A provider's `invoke` doesn't have access to the resource monitor to be able to invoke other providers, access program config, or create resources (other provider CRUD operations have this same limitation). Only `construct` or `call` can do that currently.

Ok, got it! This makes invokes totally useless for us since we need to rely on azure-native functions there. Is it possible to change that in future? It'd be nice to have a provider function that makes a bunch of calls to other provider functions to collect info and then return some result to the caller...

> I'd expect `const azureClientConfig = await azAuth.getClientConfig();` to work inside `construct`. Is it the case that it used to work with 3.121.0, but no longer works using 3.130.0?

Yes, it used to work in 3.121.0 Doesn't work in 3.122.0 and 3.130.0. Didn't check all the versions in between :)

**Comment by Frassle** [comment]
> It'd be nice to have a provider function that makes a bunch of calls to other provider functions to collect info and then return some result to the caller...

That's what `Call` is for, it's similar to `Invoke` but with access to the resource monitor.

**Comment by justinvp** [comment]
Right now the use of `Call` is limited to resource _methods_. Depending on how you're modeling things, you may be able to expose the functionality you need as a method of a component, which can be done today.

Support for using `Call` to implement a function (rather than `Invoke`) is tracked by https://github.com/pulumi/pulumi/issues/7583.

**Comment by 4ux-nbIx** [comment]
> Right now the use of `Call` is limited to resource _methods_. Depending on how you're modeling things, you may be able to expose the functionality you need as a method of a component, which can be done today.
> 
> Support for using `Call` to implement a function (rather than `Invoke`) is tracked by [#7583](https://github.com/pulumi/pulumi/issues/7583).

Yeah, already tried them in our provider. Works great if we have a resource instance! But the use case we are looking to implement is a static utility function not related to any existing resource and it needs to call similar utility functions like azure-native's getClientConfig() inside. 

Anyway, it's not a very important use case for us. And I have a couple of ideas how to work around the limitation.

Right now it'd be nice to fix construct calls :)

**Comment by justinvp** [comment]
@4ux-nbIx, I just tried this out and am unable to reproduce the error you're seeing. One thing that's suspicious in the error message, it looks like you have an `@pulumi/pulumi` package nested within the `@pulumi/azure-native` package in your `node_modules`. I wonder if this is the root of the problem.

```
at C:\snapshot\pulumi-resource-azure-relay\node_modules\@pulumi\azure-native\node_modules\@pulumi\pulumi
```

Could you try deleting `C:\snapshot\pulumi-resource-azure-relay\node_modules` and the associated `package-lock.json` (or equivalent) file, and re-running `npm install`?

Ideally, you'd end up with:

```
C:\snapshot\pulumi-resource-azure-relay\node_modules\@pulumi\azure-native
C:\snapshot\pulumi-resource-azure-relay\node_modules\@pulumi\pulumi
```

without

```
C:\snapshot\pulumi-resource-azure-relay\node_modules\@pulumi\azure-native\node_modules\@pulumi\pulumi
```

**Comment by 4ux-nbIx** [comment]
@justinvp you were absolutely right! Sorry for not noticing this myself first :)

I did do a full cleanup while trying different versions of pulumi. But! I was missing package-lock.json/yarn.lock cleanup step in our scripts. I also removed @pulumi/eslint package dependency just in case. 

Did a full cleanup and it worked!

Now we've tested it everywhere and all is good. Unfortunately I do not know what was the actual culprit: not removing package-lock.json/yarn.lock or @pulumi/eslint dependency :(

Anyway, thanks for support and explanations! And sorry for bothering you for nothing :)
